### PR TITLE
Add a script to scan source code for basic stats.

### PR DESCRIPTION
**Pull request by chandlerc**
This is a rough script that uses regexes to do a simple scan of source code and extract some basic source code statistics. Things like column width, comment line density, identifier lengths and densities.

After scanning, it prints out both raw stats and in a few cases renders a quick histogram to the terminal to help visualize a relevant distribution.

I threw this together pretty quickly, and this is an area of Python I have very limited familiarity with, so happy to have any suggestions for how to better approach this.

### CODE REVIEW

File: scripts/source_stats.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        if width in self.identifier_widths:
+            self.identifier_widths[width] += 1
+        else:
+            self.identifier_widths[width] = 1
+
+    def add_ids_per_line(self, num_ids: int) -> None:
+        if num_ids in self.ids_per_line:
+            self.ids_per_line[num_ids] += 1
+        else:
+            self.ids_per_line[num_ids] = 1
```
Comment (josh11b) [inline scripts/source_stats.py]: Usual way to do this is using https://docs.python.org/3/library/stdtypes.html#dict.get , https://docs.python.org/3/library/collections.html#collections.Counter , or `setdefault` like you do below.

File: scripts/source_stats.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+    def accumulate(self, other: _Stats) -> None:
+        self.lines += other.lines
+        self.blank_lines += other.blank_lines
+        self.empty_comment_lines += other.empty_comment_lines
+        self.comment_lines += other.comment_lines
+        for width, count in other.comment_line_widths.items():
+            self.comment_line_widths.setdefault(width, 0)
+            self.comment_line_widths[width] += count
```
Comment (josh11b) [inline scripts/source_stats.py]: Could call `self.add_comment_line_width(width, count)` if you gave that method an optional `count` parameter that defaults to `1`.

File: scripts/source_stats.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    (?P<end_open_curly>{\s*(?P<open_curly_trailing_comment>//.*)?)|
+    (?P<trailing_comment>//.*)|
+    (?P<id>\b\w+\b)
+""",
+    re.X,
+)
+
+
+@dataclass
+class _Stats:
```
Comment (josh11b) [inline scripts/source_stats.py]: Using `_` prefix seems like extra noise for a script like this.

Comment (chandlerc) [reply scripts/source_stats.py]: Done.

File: scripts/source_stats.py (hunk 1 above)
Comment (chandlerc) [reply scripts/source_stats.py]: Ah, `Counter` is perfect. I somehow didn't find that when trying to figure out how to do this in Python, and ended up making a really bad version of `Counter` here, so thanks.

File: scripts/source_stats.py (hunk 2 above)
Comment (chandlerc) [reply scripts/source_stats.py]: Obviated now that I'm using `Counter`.

**Review by chandlerc — COMMENTED** [review]
PTAL!

**Review by josh11b — APPROVED** [review]

**Pull request merged**